psi4/driver/qcdb/intf_dftd3/dashparam.py

```diff
@@ -0,0 +1,608 @@
+                    dashlevel_candidate_2, dashlevel_candidate_1))
+    dashleveleff = dashlevel_candidate_1 or dashlevel_candidate_2
+
+    allowed_params = dashcoeff[dashleveleff]['default'].keys()
+
+    # << 2 >> use name_hint and/or param_tweaks to determine intended dispersion parameters
+    if name_hint is None and param_tweaks is None:
+        raise ValidationError("""Can't guess -D parameters without name_hint ({}) or param_tweaks ({})""".format(
+            name_hint, param_tweaks))
+
+    if isinstance(param_tweaks, list):
+        param_tweaks = dict(zip(allowed_params, param_tweaks))
+
+    if param_tweaks is None:
+        param_tweaks = {}
+
+    if not set(param_tweaks.keys()).issubset(allowed_params):
+        raise ValidationError('Requested keys ({}) not among allowed ({}) for dispersion level ({})'.format(
+            list(param_tweaks.keys()), list(allowed_params), dashleveleff))
+
+    disp_params.update(param_tweaks)
+
+    if disp_params.keys() != allowed_params:
+        raise ValidationError('Requested keys ({}) insufficient ({}) for dispersion level ({})'.format(
+            list(param_tweaks.keys()), list(allowed_params), dashleveleff))
+
+    # << 3 >> use final dashlevel and disp_params to determine if a defined "fctl-disp" label exists
+    # * if/elif chooses right label when some fctls have identical param sets
+    if ((name_hint is not None) and (dashcoeff_supplement is not None)
+            and (name_key in dashcoeff_supplement[dashleveleff]['definitions'])
+            and (disp_params == dashcoeff_supplement[dashleveleff]['definitions'][name_key]['params'])):
+        fctldasheff = name_key
+        citeff = dashcoeff_supplement[dashleveleff]['definitions'][name_key].get('citation', '')
+    elif name_hint not in [None, ''] and (disp_params == dashcoeff[dashleveleff]['definitions'][name_key]['params']):
+        fctldasheff = '-'.join([name_key, dashcoeff[dashleveleff]['formal'].lower()])
+        citeff = dashcoeff[dashleveleff]['definitions'][name_key].get('citation', '')
+    else:
+        if dashcoeff_supplement is not None:
+            for func, params in dashcoeff_supplement[dashleveleff]['definitions'].items():
+                if disp_params == params['params']:
+                    fctldasheff = func
+                    citeff = params.get('citation', '')
+                    break
+
+        for func, params in dashcoeff[dashleveleff]['definitions'].items():
+            if disp_params == params['params']:
+                fctldasheff = '-'.join([func, dashcoeff[dashleveleff]['formal'].lower()])
+                citeff = params.get('citation', '')
+                break
+        else:
+            fctldasheff = ''
+            citeff = ''
+
+    # TODO right now fctldasheff is empty if undefined. use '-dash' or 'custom dash' instead?
+    # TODO right now citation is empty if undefined. remove key or use None or False instead?
```

Comment:
I think leaving both empty is fine. The first one should be always checked against `get_dispersion_aliases`, and the second one is only used for print, so we get a newline at worst.

Reply:
Yes, it's fine for here, I think, too, since this is the as-much-as-can-be-validated form of the data. When it gets to the QCSchema level, the maximal-known-fields form may be preferred.